rust/lance-index/src/vector/ivf/shuffler.rs

```
@@ -95,69 +260,84 @@ pub async fn shuffle_dataset(
+                                ),
                             );
-                            assert_eq!(indices.len(), batch.num_rows() - part_ids.null_count());
-                            batch = batch.take(&indices)?;
+                            let part_ids = UInt32Array::from(part_ids);
+                            batch = batch
+                                .try_with_column(
+                                    Field::new(PART_ID_COLUMN, part_ids.data_type().clone(), true),
+                                    Arc::new(part_ids.clone()),
+                                )
+                                .expect("failed to add part id column");
+
+                            if part_ids.null_count() > 0 {
+                                info!(
+                                    "Filter out rows without valid partition IDs: null_count={}",
+                                    part_ids.null_count()
+                                );
+                                let indices = UInt32Array::from_iter(
+                                    part_ids
+                                        .iter()
+                                        .enumerate()
+                                        .filter_map(|(idx, v)| v.map(|_| idx as u32)),
+                                );
+                                assert_eq!(indices.len(), batch.num_rows() - part_ids.null_count());
+                                batch = batch.take(&indices)?;
+                            }
                         }
-                    }
 
-                    // Filter out NaNs/Infs
-                    batch = nan_filter.transform(&batch)?;
-
-                    ivf.transform(&batch)
+                        // Filter out NaNs/Infs
+                        let minibatch_size = std::env::var("LANCE_SHUFFLE_BATCH_SIZE")
+                            .unwrap_or("64".to_string())
+                            .parse::<usize>()
+                            .unwrap_or(64);
```

If someone does specify this but there is a parsing error, I think we should either log or error, rather than silently fall back to the default.